workers/ohsome_quality_analyst/ohsome/client.py

```diff
@@ -97,61 +122,89 @@ async def query_ohsome_api(url: str, data: dict, headers: dict = {}) -> dict:
 
 
 async def get_latest_ohsome_timestamp() -> datetime.datetime:
-    """Get unix timestamp of ohsome from ohsome api."""
-    url = "https://api.ohsome.org/v1/metadata"
-    headers = {"user-agent": USER_AGENT}
+    """Get latest unix timestamp from the ohsome API."""
     async with httpx.AsyncClient() as client:
-        resp = await client.get(url, headers=headers)
-    timestamp_str = str(resp.json()["extractRegion"]["temporalExtent"]["toTimestamp"])
-    timestamp = datetime.datetime.strptime(timestamp_str, "%Y-%m-%dT%H:%MZ")
-    return timestamp
+        resp = await client.get(
+            url=OHSOME_API.rstrip("/") + "/metadata",
+            headers={"user-agent": USER_AGENT},
+        )
+    strtime = resp.json()["extractRegion"]["temporalExtent"]["toTimestamp"]
+    return datetime.datetime.strptime(strtime, "%Y-%m-%dT%H:%MZ")
 
 
 def build_url(
-    layer,
-    endpoint: Optional[str] = None,
+    layer: Layer,
     ratio: bool = False,
-) -> str:
-    """Build endpoint URL of ohsome API."""
-    ohsome_api = OHSOME_API.rstrip("/")
-    if endpoint is None:
-        endpoint = layer.endpoint
-    url = ohsome_api + "/" + endpoint.rstrip("/")
+    group_by: bool = False,
+    contributions: bool = False,
+):
+    if contributions:
+        return OHSOME_API.rstrip("/") + "/contributions/latest/count"
+    url = OHSOME_API.rstrip("/") + "/" + layer.endpoint.rstrip("/")
     if ratio:
-        return url + "/" + "ratio"
+        url += "/ratio"
+    if group_by:
+        url += "/groupBy/boundary"
```

Comment:
Not all combinations are valid. Where do you validate invalid requests?

Reply:
There is no validation in code. The only validation that takes place is to make sure the manual defined layers are correct. 

The ratio endpoint is only used by the TagRatio Indicator. The TagRatio Indicator does only work for certain layer manually defined (see `definitions.py` and `INDICATOR_LAYER` variable). Those layer have to have a second filter element.

Reply:
https://github.com/GIScience/ohsome-quality-analyst/issues/300